### ISSUE TITLE
`softmax` to handle dimensions comprised of `-inf`

### ISSUE THREAD

**Issue by nikitaved** [pytorch/pytorch#106545]
### 🚀 The feature, motivation and pitch

Currently `softmax` handles `-inf` in most of the cases, for example:

```python
>>> x = torch.tensor([-float('inf'), 1])
>>> x.softmax(-1)
tensor([0., 1.])
>>> x.cuda().softmax(-1)
tensor([0., 1.], device='cuda:0')
```

However, once the normalized dimension is comprised of solely `-inf`, there is an issue:
```python
>>> x = torch.tensor([-float('inf')] * 2)
>>> x.softmax(-1)
tensor([nan, nan])
>>> x.cuda().softmax(-1)
tensor([nan, nan], device='cuda:0')
```

I understand we get it because `inf - inf = nan`, but why not assume that `softmax` is computed in the limit?
In such a case `softmax` should return a uniform distribution. This case is quite easy to detect since `max == -inf` will imply a uniform distribution.

Issues for more context:
https://github.com/pytorch/pytorch/issues/25110
https://github.com/pytorch/pytorch/issues/103749
https://github.com/pytorch/pytorch/issues/103963

**Comment by mikaylagawarecki** [comment]
Hey @nikitaved, since `-inf` is usually used for masking, a uniform distribution might not be what we want here as that gives a uniform weight to everything --> attending to everything, which is wrong.

You might be able to workaround this by writing a custom autograd.Function (or CustomOp) that has the desired behavior and using it in conjunction with torch.compile.
